connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -0,0 +1,537 @@
+                    String recoveredWorker = candidateWorkersForReassignment.iterator().next();
+                    WorkerLoad workerLoad = completeWorkerAssignment.stream()
+                            .filter(w -> recoveredWorker.equals(w.worker()))
+                            .findAny()
+                            .orElse(null);
+
+                    if (workerLoad != null) {
+                        lostAssignments.connectors().forEach(c -> workerLoad.assign(c));
+                        lostAssignments.tasks().forEach(t -> workerLoad.assign(t));
+                    }
+                } else {
+                    newSubmissions.connectors().addAll(lostAssignments.connectors());
+                    newSubmissions.tasks().addAll(lostAssignments.tasks());
+                }
+                scheduledRebalance = 0;
+            } else {
+                if (now < scheduledRebalance) {
+                    // a delayed rebalance is in progress, but it's not yet time to reassign
+                    // unaccounted resources
+                    delay = calculateDelay(now);
+                } else {
+                    // This means scheduledRebalance == 0
+                    candidateWorkersForReassignment.addAll(completeWorkerAssignment.stream()
+                            .filter(WorkerLoad::isEmpty)
+                            .map(WorkerLoad::worker)
+                            .collect(Collectors.toSet()));
+                    // we can also extra the current minimum delay from the group, to make
+                    // independent of consecutive leader failures.
+                    delay = maxDelay;
+                }
+                scheduledRebalance = now + delay;
+            }
+        }
+    }
+
+    /**
+     * TODO: describe algorithm
+     *
+     * @param activeAssignments
+     * @param completeWorkerAssignment
+     * @return
+     */
+    private Map<String, ConnectorsAndTasks> performTaskRevocation(ConnectorsAndTasks activeAssignments,
+                                                                  Collection<WorkerLoad> completeWorkerAssignment) {
+        int totalActiveConnectorsNum = activeAssignments.connectors().size();
+        int totalActiveTasksNum = activeAssignments.tasks().size();
+        Collection<WorkerLoad> existingWorkers = completeWorkerAssignment.stream()
+                .filter(wl -> wl.size() > 0)
+                .collect(Collectors.toList());
+        int existingWorkersNum = existingWorkers.size();
+        int totalWorkersNum = completeWorkerAssignment.size();
+        int newWorkersNum = totalWorkersNum - existingWorkersNum;
+
+        Map<String, ConnectorsAndTasks> revoking = new HashMap<>();
+        if (!(newWorkersNum > 0 && existingWorkersNum > 0)) {
+            log.debug("Business as usual existing {} new {} total {}", existingWorkersNum,
+                    newWorkersNum, totalWorkersNum
+            );
+
+            if (log.isDebugEnabled()) {
```

Comment:
Why the explicit check for log level here?

Reply:
My rule of thumb is that I include in a if statement only the debug statements that would result in eager (and potential wasteful) computation of print arguments. The rest of the debug logging is kept outside the if branch, which will allow us to remove the if branch altogether if the line that requires it is removed. Otherwise, we risk keeping redundant `if`s around.